crates/qasm2/src/error.rs

```diff
@@ -0,0 +1,85 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::import_exception;
+
+use crate::lex::Token;
+
+pub struct Position<'a> {
+    filename: &'a std::ffi::OsStr,
+    line: usize,
+    col: usize,
+}
+
+impl<'a> Position<'a> {
+    pub fn new(filename: &'a std::ffi::OsStr, line: usize, col: usize) -> Self {
+        Self {
+            filename,
+            line,
+            col,
+        }
+    }
+}
+
+impl<'a> std::fmt::Display for &Position<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{},{}",
+            self.filename.to_string_lossy(),
+            self.line,
+            self.col
+        )
+    }
+}
+
+/// Create an error message that includes span data from the given [token][Token].  The base of the
+/// message is `message`, and `filename` is the file the triggering OpenQASM 2 code came from.  For
+/// string inputs, this can be a placeholder.
+pub fn message_generic(position: Option<&Position>, message: &str) -> String {
+    if let Some(position) = position {
+        format!("{}: {}", position, message)
+    } else {
+        message.to_owned()
+    }
+}
+
+/// Shorthand form for creating an error message when a particular type of token was required, but
+/// something else was `received`.
+pub fn message_incorrect_requirement(
+    required: &str,
+    received: &Token,
+    filename: &std::ffi::OsStr,
+) -> String {
+    message_generic(
+        Some(&Position::new(filename, received.line, received.col)),
+        &format!(
+            "needed {}, but instead saw {}",
+            required,
+            received.ttype.describe()
+        ),
+    )
+}
+
+/// Shorthand form for creating an error message when a particular type of token was required, but
+/// the input ended unexpectedly.  The `owner` [Token] is whatever the token was that caused us to
+/// know something else was required.
+pub fn message_bad_eof(position: Option<&Position>, required: &str) -> String {
+    message_generic(
+        position,
+        &format!("unexpected end-of-file when expecting to see {}", required),
+    )
+}
+
+// We define the exception in Python space so it can inherit from QiskitError; it's easier to do
+// that from Python and wrap rather than also needing to import QiskitError to Rust to wrap.
+import_exception!(qiskit.qasm2.exceptions, QASM2ParseError);
```

Comment:
You define 3 exceptions in Python land but only import 1 here

Reply:
That's expected - the Rust code only needs to emit the concrete `QASM2ParseError`, never `QASM2ExportError`.  `QASM2Error` is more structural than a concrete class, but it's in place for a follow-up PR where I'll alias the existing `QasmError` to `QASM2Error` as part of the complete removal of the old `qiskit.qasm` namespace.  I actually only meant to define two exceptions in Python space in this PR - the `QASM2ExportError` one doesn't need to be in this PR, I just missed it when rebasing.

Reply:
Reading it again, the comment is kind of an artifact of how I vendored the package - over on https://github.com/jakelishman/qiskit-qasm2, I just have a single exception defined in Rust space, so when I wrote the comment after fixing things up into Qiskit structure, I had that historical background in mind.